ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs

```diff
@@ -172,98 +191,141 @@ tickHeaderState :: ConsensusProtocol (BlockProtocol blk)
+data HeaderStateHistory blk = HeaderStateHistory {
+      -- | The last, right-most, element in the sequence contains the current
+      -- state, see 'headerStateHistoryCurrent'.
+      headerStateHistorySnapshots :: !(StrictSeq (HeaderState blk))
 
-    trim :: StrictSeq (AnnTip blk) -> Maybe (AnnTip blk, StrictSeq (AnnTip blk))
-    trim (newAnchor :<| trimmed) | Seq.length trimmed >= fromIntegral k =
-        Just (newAnchor, trimmed)
-    trim _otherwise = Nothing
+      -- | Header state /before/ the oldest in 'headerStateHistorySnapshots'.
+      -- This is the oldest tip we can roll back to.
+    , headerStateHistoryAnchor    :: !(HeaderState blk)
+    }
+  deriving (Generic)
 
 deriving instance (BlockSupportsProtocol blk, HasAnnTip blk)
-                => Show (HeaderState blk)
+                => Eq (HeaderStateHistory blk)
 deriving instance (BlockSupportsProtocol blk, HasAnnTip blk)
-                => NoUnexpectedThunks (HeaderState blk)
-deriving instance ( BlockSupportsProtocol blk
-                  , HasAnnTip blk
-                  , Eq (ChainDepState (BlockProtocol blk))
-                  ) => Eq (HeaderState blk)
-
-genesisHeaderState :: ChainDepState (BlockProtocol blk) -> HeaderState blk
-genesisHeaderState state = HeaderState state Seq.Empty Origin
+                => Show (HeaderStateHistory blk)
+deriving instance (BlockSupportsProtocol blk, HasAnnTip blk)
+                => NoUnexpectedThunks (HeaderStateHistory blk)
+
+headerStateHistoryCurrent :: HeaderStateHistory blk -> HeaderState blk
+headerStateHistoryCurrent HeaderStateHistory {..} =
+    case headerStateHistorySnapshots of
+      _ :|> current -> current
+      Empty         -> headerStateHistoryAnchor
+
+headerStateHistoryCurrentPoint ::
+     HasAnnTip blk
+  => HeaderStateHistory blk -> Point blk
+headerStateHistoryCurrentPoint =
+      withOriginRealPointToPoint
+    . fmap annTipRealPoint
+    . headerStateTip
+    . headerStateHistoryCurrent
+
+-- | Append a 'HeaderState' to the history.
+headerStateHistoryPush ::
+     forall blk.
+     HeaderState blk
+  -> HeaderStateHistory blk
+  -> HeaderStateHistory blk
+headerStateHistoryPush headerState HeaderStateHistory {..} =
+    HeaderStateHistory {
+        headerStateHistorySnapshots = headerStateHistorySnapshots :|> headerState
+      , headerStateHistoryAnchor    = headerStateHistoryAnchor
+      }
 
-castHeaderState :: ( Coercible (ChainDepState (BlockProtocol blk ))
-                               (ChainDepState (BlockProtocol blk'))
-                   , TipInfo blk ~ TipInfo blk'
-                   )
-                => HeaderState blk -> HeaderState blk'
-castHeaderState HeaderState{..} = HeaderState{
-      headerStateConsensus = coerce headerStateConsensus
-    , headerStateTips      = castSeq castAnnTip $ headerStateTips
-    , headerStateAnchor    = fmap    castAnnTip $ headerStateAnchor
+-- | Trim the 'HeaderStateHistory' to the given size, dropping the oldest
```

Comment:
`headerStateHistoryTrim 5 x` will result in a `HeaderStateHistory` data structure that ultimately contains 6 header states, right? (It's +1 because I'm counting the anchor state.) That makes this comment a little ambiguous.